protocol/contracts/templegold/DaiGoldAuction.sol

```diff
@@ -0,0 +1,339 @@
+    address public override auctionStarter;
+
+    /// @notice Keep track of next epoch auction Temple Gold amount
+    uint256 public override nextAuctionGoldAmount;
+
+    /// @notice Auction duration
+    uint64 public constant AUCTION_DURATION = 1 weeks;
+    /// @notice Configuration for auction
+    AuctionConfig private auctionConfig;
+
+    constructor(
+        address _templeGold,
+        address _bidToken,
+        address _treasury,
+        address _rescuer,
+        address _executor,
+        address _auctionStarter
+    ) TempleElevatedAccess(_rescuer, _executor) {
+        templeGold = ITempleGold(_templeGold);
+        bidToken = IERC20(_bidToken);
+        treasury = _treasury;
+        auctionStarter = _auctionStarter;
+        emit AuctionStarterSet(_auctionStarter);
+    }
+
+    /**
+     * @notice Set auction configuration
+     * @param _config Auction configuration
+     */
+    function setAuctionConfig(AuctionConfig calldata _config) external override onlyElevatedAccess {
+        if (_config.auctionStartCooldown == 0
+                || _config.auctionMinimumDistributedGold == 0
+                || _config.auctionsTimeDiff == 0) 
+            { revert CommonEventsAndErrors.ExpectedNonZero(); }
+        if (!epochs[_currentEpochId].hasEnded()) { revert InvalidOperation(); }
+        auctionConfig = _config;
+
+        emit AuctionConfigSet(_currentEpochId, _config);
+    }
+
+    /**
+     * @notice Set address to trigger auction start. Zero address accepted
+     * @param _starter Auction starter
+     */
+    function setAuctionStarter(address _starter) external override onlyElevatedAccess {
+        /// @notice No zero address checks. Zero address is a valid input
+        auctionStarter = _starter;
+        if (!epochs[_currentEpochId].hasEnded()) { revert InvalidOperation(); }
+        emit AuctionStarterSet(_starter);
+    }
+
+    /**
+     * @notice Set token used for bidding
+     * @param _bidToken Bid token
+     */
+    function setBidToken(address _bidToken) external override onlyElevatedAccess {
+        if (_bidToken == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+        if (!epochs[_currentEpochId].hasEnded()) { revert InvalidOperation(); }
+        bidToken = IERC20(_bidToken);
+        emit BidTokenSet(_bidToken);
+    }
+
+    /**
+     * @notice Start auction. Auction start can be triggered by anyone if `auctionStarter` not set
+     * @dev The Temple Gold amount for the auction is fixed and set at startAuction(). 
+     * So in `startAuction()`, there is a call to `_distributeGold()` to mint and distribute TGOLD. 
+     * Any other `_distributeGold()` calls during auction is tracked for next auction use.
+     */
+    function startAuction() external override {
+        if (auctionStarter != address(0) && msg.sender != auctionStarter) { revert CommonEventsAndErrors.InvalidAccess(); }
+        EpochInfo storage prevAuctionInfo = epochs[_currentEpochId];
```

Comment:
tiny gas saving:

`_currenEpochId` is read from storage multiple times in this function. It could be cached in memory